Documentation/api-guidelines/nullability.md

```diff
@@ -0,0 +1,77 @@
+
+A code review for enabling nullability generally involves two or three passes:
+
+1.	Review all implementation changes made in the code.  Except when explicitly fixing a bug (which should be rare), the annotations employed for nullability should have zero impact on the generated IL (other than potentially some added attributes in the metadata).  The most common changes are:
+a.	Adding `?` to reference type parameters and local symbols.  These inform the compiler that nulls are allowed.  For locals, they evaporate entirely at compile time.  For parameters, they impact the [Nullable(�)] attributes emitted into the metadata by the compiler, but have no effect on the implementation IL.
+b.	Adding `!` to reference type usage.  These essentially suppress the null warning, telling the compiler to treat the expression as if it�s non-null.  These evaporate at compile-time.
+c.	Adding `Debug.Assert(reference != null);` statements.  These inform the compiler that the mentioned reference is non-null, which will cause the compiler to factor that in and have the effect of suppressing subsequent warnings on that reference (until the flow analysis suggests that could change).  As with any Debug.Assert, these evaporate at compile-time in release builds (where DEBUG isn�t defined).
+Most any other changes have the potential to change the IL, which should not be necessary for the feature.  In particular, it�s common for `?`s on dereferences to sneak in, e.g. changing `someVar.SomeMethod()` to `someVar?.SomeMethod()`; that is a change to the IL, and should only be employed when there�s an actual known bug that�s important to fix, as otherwise we�re incurring unnecessary cost.
+2.	Review the API changes explicitly made.  These are the ones that show up in the diff.  They should be reviewed to validate that they make sense from a contract perspective.  Do we expect/allow nulls everywhere a parameter reference type was augmented with `?`?  Do we potentially return nulls everywhere a return type was augmented with `?`?  Was anything else changed that could be an accidental breaking change (e.g. a value type parameter getting annotated to become a Nullable<T> instead of a T)?
```

Comment:
```suggestion
2.	**Review the API changes explicitly made.**  These are the ones that show up in the diff.  They should be reviewed to validate that they make sense from a contract perspective.  Do we expect/allow nulls everywhere a parameter reference type was augmented with `?`?  Do we potentially return nulls everywhere a return type was augmented with `?`?  Was anything else changed that could be an accidental breaking change (e.g. a value type parameter getting annotated to become a Nullable<T> instead of a T)?
```